Reduce Concurrent Tests

anissa111 opened the pull request:
Addresses #473

Summary of changes:
- adds separate `ci-nightly.yml` workflow that runs on all `os` and `python-version`s at 1 AM nightly
- modifies `ci.yml` to cancel previous workflows and to run `macos-latest` on oldest and newest versions of python we support (3.7 and 3.10) to help concurrency limit (all versions run nightly)

anissa111 commented:
Note that previous tests on a branch can now be automatically cancelled as shown last commit

michaelavs reviewed (approved):
Approving as it looks like all checks will pass with these changes (I assume the sphinx one would/will pass if given enough time)

anissa111 commented:
That... actually shouldn't be there. Huh. I wonder if it's because it's not in main? If you click on the actions for the PR, it shows them all as passing. This PR specifically removed macos-latest 3.9

hCraker commented:
> Approving as it looks like all checks will pass with these changes (I assume the sphinx one would/will pass if given enough time)

@michaelavs Do you think we should wait for it to pass before merging, or is it worth risking? I highly doubt a preemptive merge will break everything (I could regret saying that)

anissa111 commented:
@hCraker We don't have the permissions to do that anymore haha

hCraker commented:
> @hCraker We don't have the permissions to do that anymore haha

Oh darn I didn't even notice we lost that since I used it so rarely

anissa111 commented:
@erogluorhan We think this might need your admin privileges. The checks on the PR start with just those we've made in the PR and then adds at least one extra expected required test (from main?) that is one of the tests that we've removed in this PR and on it indefinitely, which blocks merging.

Here is a screenshot of the tests right after I closed and reopened the PR, which lists the expected number of tests
![image](https://user-images.githubusercontent.com/38434768/179311570-003af005-ad94-4a49-a5dc-4761576060a5.png)

Here is a screenshot taken a few minutes afterward that shows that two extra tests have been added. I think one is `docs/readthedocs.org:geocat-examples` and the other is `macos-latest 3.9`
![image1](https://user-images.githubusercontent.com/38434768/179311734-020cc812-5320-460c-842d-3b6d90fd1677.png)

I am unclear why only `macos-latest 3.9` is acting like this and not `macos-latest 3.8`, which was also removed in this PR, but I suspect it is because 3.8 tests were added relatively recently and might not be part of the branch protection rules for main, but we don't have the permissions to see those, either.

If you look at the actions tab, it shows [all tests pass](https://github.com/NCAR/GeoCAT-examples/actions/runs/2678655490) and lists the expected combinations of os and python version.

When you have a minute, could you check this out and maybe lend this PR your magic admin privileges?

erogluorhan commented:
> > Approving as it looks like all checks will pass with these changes (I assume the sphinx one would/will pass if given enough time)
> 
> @michaelavs Do you think we should wait for it to pass before merging, or is it worth risking? I highly doubt a preemptive merge will break everything (I could regret saying that)

It should (almost) never be worth risking the `main` branch without seeing all of them pass.

erogluorhan commented:
> @hCraker We don't have the permissions to do that anymore haha

As soon as we have package leaders listed, let us give them the admin privileges, too (but, with they swearing not to risk the `main` branch in any circumstances haha)

erogluorhan commented:
@anissa111 sorry if I am too late, but I figured why only 3.9 was acting like that. Here is what the current state of our branch protection rule was (3.8 was not listed there):

<img width="732" alt="image" src="https://user-images.githubusercontent.com/32553057/179327603-62433ac8-0bf8-4460-a26a-4f938be6e9e6.png">

anissa111 commented:
@erogluorhan No problem! I figured it was something like that. I suggest removing `macos-latest 3.9` and adding `macos-latest 3.10`, `ubuntu-latest 3.8`, and `ubuntu-latest 3.10`

anissa111 commented:
@erogluorhan Great, thank you!!

erogluorhan commented on .github/workflows/ci.yml:
```diff
@@ -16,8 +18,13 @@ jobs:
       fail-fast: false
       matrix:
-        os: [ "ubuntu-latest", "macos-latest"]
+        os: [ "ubuntu-latest"]
```
Instead of removing macos-latest at all, I'd do ubuntu and macos but only for versions 3.7 and 3.10 for this pull-request CI

erogluorhan commented on .github/workflows/ci-nightly.yml:
```diff
@@ -0,0 +1,37 @@
+name: CI
```
I'd name this CI Nightly

anissa111 replied:
I agree! See https://github.com/NCAR/GeoCAT-examples/pull/475
